### ISSUE TITLE
`vue/require-valid-default-prop` must be a function when using intersection

### ISSUE THREAD

**Issue by krystof-k** [vuejs/eslint-plugin-vue#2060]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 8.5.0
- **eslint-plugin-vue version:** 9.0.0
- **Node version:** 18.7.0
- **Operating System:** macOS Ventura

**Please show your full configuration:**

<details>
<summary>.eslintrc.cjs</summary>

```json5
/* eslint-env node */
require('@rushstack/eslint-patch/modern-module-resolution');

module.exports = {
  root: true,
  extends: [
    'plugin:vue/vue3-recommended',
    'eslint:recommended',
    '@vue/eslint-config-typescript/recommended',
    '@vue/eslint-config-prettier',
  ],
  overrides: [
    {
      files: ['cypress/e2e/**.{cy,spec}.{js,ts,jsx,tsx}'],
      extends: ['plugin:cypress/recommended'],
    },
  ],
};
```

</details>

**What did you do?**	

```vue
<script setup lang="ts">
export type Dummy = 'one' | 'two' | 'three';

export type Restricted = ('one' | 'two' | 'three') & ('one' | 'three');

export interface Props {
  dummy?: Dummy;
  restricted?: Restricted;
}

withDefaults(defineProps<Props>(), {
  dummy: 'one',
  restricted: 'one',
});
</script>
```

**What did you expect to happen?**

No lint errors.

**What actually happened?**

I'm getting

```
Type of the default value for 'restricted' prop must be a function.
```

for the `restricted` prop.

**Repository to reproduce this issue**

[Here](https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyNV8tu5DYQ/BVBh9gLmDCQo/eRHPYDgpx9oajWiGuKVEhqHnH87ymSI0qUtEEunlFVq7tZ/eD4vRampfql/uKElaOvHPlprBTXp6+vtXev9bdXTdfRWF/520jV92kYbtXX6sFoeqj+qR78xaTP3hI9fH7Vpf2f5LyVwlOLlx5/8tan6pc1l7C1K6k92Y4Lqv6wZnTV+6uuqjak8ttLSulzQGwOBniJDO4jOLtI33+njk/Ku8eWOqkpuvsS/357/PS0cvySjvhU+l3Qj5Dgl+ckG0Sqn2o7KXL1y3t9nuhZmGEg7VkrLQkvzxD516fI/HBXNjly7Myti6gwGgEm4Y1lbhrJRrTD0/1toyNyIg8dmEWNbEK0YdzdtGCjNYN0xOhKYoKfmRUcYEtCccuDH5cJxZ3Dy06esivkPHJLTNOJ/U3WLLhud6ZI+QjjOHOwlznrxHhrFDI/0XUGW2qm04ns8qzIU1AlI9CCcXtyBZAyH2hoyJYMKRxWdgX2Rre1kZKCI0iQZUZpGP2teGCih1wiaB2DleTIQxXy2ei6kYGu3nLWGKOI6xDJl4yjQc5Ix5VCv5vp1GdoQjlLjydlGq42oBzCaGxBrZH0Ub2lPnMl21SCMaMWzxOM2aWXntyIGZs5ZSCz6dBaJKRbVRONhqO1Up9+JtQgr9Sy6A09hubxvMmkpgtzt6ExKiNGG+FxQiQuB3ySE3zMiZjmB2RUKnuIxvMDOt+bsGtYM0nl5XJkS0kIWgCc/Z7WjDlS3UZEdzRmruetubBlFTDNh5UbDI4LzWr50m++l441hEGm1VyDmDTWz/KAglB0OWA1SYW9tHCWOERu1ApyvCOG7QVJbhsUk8vXg3eHzRhACAtXUqNwC49N1DLFG1rkvYN5PQUIMhHaoeHizVJHlvTSKDOJyRL9FixLGVZw/G7prwm7jd0kqTYiU5hdp3nKPTVrKKvp8uaM8rCLsW1cVdjEWDZLGYJJWIjWolBhjOIAoPPZJWc228SlCSb4mYL0aCMUyMvSl+ilauP6QqiC2L8X3LXc87VZuYLW4LnYVgtzX1DYMVY2iFC8ma7DECjdOyxe12uLYfJoAExmyKt4NzRoIAjy2DWBekIPFo6BUS0ZR/ZM/6l2ttmeMhO7TGLSCcbEp5tvslSYYK8UcbfCOtli33cd5sb9jzp6rG4VdEWWJX71iMShG64u0RdJ3OcgJ+EOyDwkM4yabnO5MSXfCDeW7027NT6jBeJY4FtXFmYeoZz8cVOc4ynw0ftBMaOXhLPVPG2LntLtyJBtnDi2+j2ys8Lkt+VyXLPxDIXImYkD3aafXzHWyiT4Oqri7cAGt6d3yR1fJ5jExPHpihspwyluVi428IZMPwWT42OqbOFE6SC6l+Jtg+dqWWO2aZxZI1cjNoNCGb51s98Pa3wHrltnxkI/7MADhwP+BdgbGn0AiX3k0e4x15vLHlQHeoTWBfjxVMcLFIeofw89eN9v5HAb+uc79/Ev+JdiVA==).

**Comment by krystof-k** [comment]
A related issue is also in `@vue/compiler-sfc` which is already fixed (#7394)!

**Comment by waynzh** [comment]
Perhaps this issue has been fixed. I can't reproduce [this one](https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyNV8tu5DYQ/BVBh9gLmDCQo/eRHPYDgpx9oajWiGuKVEhqHnH87ymSI0qUtEEunlFVq7tZ/eD4vRampfql/uKElaOvHPlprBTXp6+vtXev9bdXTdfRWF/520jV92kYbtXX6sFoeqj+qR78xaTP3hI9fH7Vpf2f5LyVwlOLlx5/8tan6pc1l7C1K6k92Y4Lqv6wZnTV+6uuqjak8ttLSulzQGwOBniJDO4jOLtI33+njk/Ku8eWOqkpuvsS/357/PS0cvySjvhU+l3Qj5Dgl+ckG0Sqn2o7KXL1y3t9nuhZmGEg7VkrLQkvzxD516fI/HBXNjly7Myti6gwGgEm4Y1lbhrJRrTD0/1toyNyIg8dmEWNbEK0YdzdtGCjNYN0xOhKYoKfmRUcYEtCccuDH5cJxZ3Dy06esivkPHJLTNOJ/U3WLLhud6ZI+QjjOHOwlznrxHhrFDI/0XUGW2qm04ns8qzIU1AlI9CCcXtyBZAyH2hoyJYMKRxWdgX2Rre1kZKCI0iQZUZpGP2teGCih1wiaB2DleTIQxXy2ei6kYGu3nLWGKOI6xDJl4yjQc5Ix5VCv5vp1GdoQjlLjydlGq42oBzCaGxBrZH0Ub2lPnMl21SCMaMWzxOM2aWXntyIGZs5ZSCz6dBaJKRbVRONhqO1Up9+JtQgr9Sy6A09hubxvMmkpgtzt6ExKiNGG+FxQiQuB3ySE3zMiZjmB2RUKnuIxvMDOt+bsGtYM0nl5XJkS0kIWgCc/Z7WjDlS3UZEdzRmruetubBlFTDNh5UbDI4LzWr50m++l441hEGm1VyDmDTWz/KAglB0OWA1SYW9tHCWOERu1ApyvCOG7QVJbhsUk8vXg3eHzRhACAtXUqNwC49N1DLFG1rkvYN5PQUIMhHaoeHizVJHlvTSKDOJyRL9FixLGVZw/G7prwm7jd0kqTYiU5hdp3nKPTVrKKvp8uaM8rCLsW1cVdjEWDZLGYJJWIjWolBhjOIAoPPZJWc228SlCSb4mYL0aCMUyMvSl+ilauP6QqiC2L8X3LXc87VZuYLW4LnYVgtzX1DYMVY2iFC8ma7DECjdOyxe12uLYfJoAExmyKt4NzRoIAjy2DWBekIPFo6BUS0ZR/ZM/6l2ttmeMhO7TGLSCcbEp5tvslSYYK8UcbfCOtli33cd5sb9jzp6rG4VdEWWJX71iMShG64u0RdJ3OcgJ+EOyDwkM4yabnO5MSXfCDeW7027NT6jBeJY4FtXFmYeoZz8cVOc4ynw0ftBMaOXhLPVPG2LntLtyJBtnDi2+j2ys8Lkt+VyXLPxDIXImYkD3aafXzHWyiT4Oqri7cAGt6d3yR1fJ5jExPHpihspwyluVi428IZMPwWT42OqbOFE6SC6l+Jtg+dqWWO2aZxZI1cjNoNCGb51s98Pa3wHrltnxkI/7MADhwP+BdgbGn0AiX3k0e4x15vLHlQHeoTWBfjxVMcLFIeofw89eN9v5HAb+uc79/Ev+JdiVA==). 🤔

**Comment by FloEdelmann** [comment]
Looks like it :slightly_smiling_face: 
I'll close this issue for now. Feel free to comment or open a new issue if the problem persists.

**Comment by krystof-k** [comment]
Yeah, me neither anymore, looking fine now.
